creusot/tests/should_succeed/sparse_array.rs

```diff
@@ -119,7 +119,7 @@ impl<T> Sparse<T> {
         let index = self.idx[i];
         if !(index < self.n && self.back[index] == i) {
             // the hard assertion!
-            proof_assert!( self.lemma_permutation(@i); true );
+            ghost!( Self::lemma_permutation );
```

Comment:
Is there any way we could actually call `lemma_permutation` with valid arugments here?

Reply:
Well we shouldn't be able to use `forall` or `exists` either, in fact the *only* syntax we would want to admit is `@`.

Reply:
You can't use `forall` or `exists` in why3 ghost code either afaik, it's just normal why3 code with the erasure property. That's also the reason why I chose to keep borrow checking for `ghost!` and ties into the broader discussion we have been having about having tiers of ghost code.

Reply:
> On the other hand, there is something we allow that should be forbidden: unrestricted use of logic functions are dangerous since they can themselve use prophecies.

They can't meaningfully use them since they can't perform any mutation or assumptions, and we cannot mutate borrows in ghost code. 

I do agree that we need to actually define the semantics of ghost code carefully though.

Reply:
> They can't meaningfully use them since they can't perform any mutation or assumptions, and we cannot mutate borrows in ghost code.

Actually I think they can be used for unsoundness: imagine you define a logic function which returns the prophecy of a borrow. Then if you call this in ghost code, you effectively have access to `^`. So you can create a causality loop in ghost code (for a mutable borrow of a ghost variable, for example).

Reply:
how do you close the loop?

Reply:
Something like this:

```
let x = &mut ghost! { 1 };
*x = ghost! { ^x + 1 };
```

Then , if you resolve x at this point, you assume x = x+1. Boom.

Reply:
You can't write to mutable borrows in ghost code.

Reply:
The point being: for any kind of information:

(a) either you decide that it can depend on prophetic information, and then it should not be allowed to take borrows of this information (i.e., the thing is in the clairvoyant monad)
(b) or you decide that it only depends on the operational semantics, and then borrows are allowed.

Before introducing this kind of ghost code, the situation was clear: all pearlite code was (a) and all compiled code was (b). Now that we have ghost code, we have to decide whether it belongs to (a) or (b). To me, ghost code can only be really useful if it is part of (b), because then you can place ghost information in structs etc. So then you need to make sure that you cannot use prophetic information in ghost code.

Reply:
> You can't write to mutable borrows (modulo a bug) in ghost code.

I'm not doing this. I am writing in actual code to a variable of type `Ghost`.

Reply:
> I'm not doing this. I am writing in actual code to a variable of type Ghost.

Oh that is mind breaking.

Reply:
Let's discuss this in #436